### PR TITLE
Dr/gpu

### PR DESCRIPTION
**Pull request by ravil-mobile**
Added a sketch of DR/gpu. 

* Refactored some classes
* adapted CMake

### CODE REVIEW

File: src/DynamicRupture/Factory.cpp [hunk 1]
```diff
@@ -47,7 +47,11 @@ Products NoFaultFactory::produce() {
 Products LinearSlipWeakeningFactory::produce() {
   return {std::make_unique<seissol::initializers::LTS_LinearSlipWeakening>(),
           std::make_unique<initializers::LinearSlipWeakeningInitializer>(drParameters),
+#ifndef ACL_DEVICE_OFFLOAD
```
Comment (sebwolf-de) [inline src/DynamicRupture/Factory.cpp:50]: How about this idea:
```
#include <iostream>

namespace friction_law::cpu {
    void write() {
        std::cout << "cpu" << std::endl;
    }
}
namespace friction_law::gpu {
    void write() {
        std::cout << "gpu" << std::endl;
    }
}

#ifdef ACL_DEVICE_OFFLOAD
namespace fl = friction_law::gpu;
#else
namespace fl = friction_law::cpu;
#endif

int main() {
    fl::write();
}
```
Then we only need one `ifdef`

File: src/DynamicRupture/FrictionLaws/FrictionSolver.cpp [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+  using QInterpolatedShapeT =
+      real(*)[CONVERGENCE_ORDER][misc::numQuantities][misc::numPaddedPoints];
+  auto* qIPlus = (reinterpret_cast<QInterpolatedShapeT>(qInterpolatedPlus))[ltsFace];
+  auto* qIMinus = (reinterpret_cast<QInterpolatedShapeT>(qInterpolatedMinus))[ltsFace];
+
+  for (unsigned o = 0; o < CONVERGENCE_ORDER; ++o) {
+
+#ifdef ACL_DEVICE_OFFLOAD
+#pragma omp loop bind(parallel)
+#endif // ACL_DEVICE_OFFLOAD
```
Comment (sebwolf-de) [inline src/DynamicRupture/FrictionLaws/FrictionSolver.cpp:72]: Shouldn't there be a 
```
#else
#pragma omp parallel for simd
```
?

File: src/DynamicRupture/FrictionLaws/FrictionSolver.cpp [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+                                                          double timeWeights[CONVERGENCE_ORDER],
+                                                          unsigned int ltsFace) {
+  // this initialization of the kernel could be moved to the initializer
+  // set inputParam could be extendent for this (or create own function)
+  // the kernel then could be a class attribute and following values are only set once
+  //(but be careful of race conditions since this is computed in parallel for each face!!)
+
+  // set imposed state to zero
+#ifdef ACL_DEVICE_OFFLOAD
+#pragma omp loop bind(parallel)
```
Comment (sebwolf-de) [inline src/DynamicRupture/FrictionLaws/FrictionSolver.cpp:106]: Also here: do we need an else for cpu only excution?

File: src/DynamicRupture/FrictionLaws/FrictionSolver.cpp [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+  auto* qIPlus = (reinterpret_cast<QInterpolatedShapeT>(qInterpolatedPlus))[ltsFace];
+  auto* qIMinus = (reinterpret_cast<QInterpolatedShapeT>(qInterpolatedMinus))[ltsFace];
+
+  for (unsigned o = 0; o < CONVERGENCE_ORDER; ++o) {
+    auto weight = timeWeights[o];
+
+#ifdef ACL_DEVICE_OFFLOAD
+#pragma omp loop bind(parallel)
+#else
+#pragma omp simd
```
Comment (sebwolf-de) [inline src/DynamicRupture/FrictionLaws/FrictionSolver.cpp:133]: Why no `omp parallel for simd`?

File: src/DynamicRupture/FrictionLaws/GpuImpl/LinearSlipWeakening.cpp [hunk 5]
```diff
@@ -0,0 +1,207 @@
+#include "DynamicRupture/FrictionLaws/GpuImpl/LinearSlipWeakening.h"
```
Comment (sebwolf-de) [inline src/DynamicRupture/FrictionLaws/GpuImpl/LinearSlipWeakening.cpp:1]: Is this file just a copy from the `cpu` version?

File: src/DynamicRupture/FrictionLaws/FrictionSolver.cpp (hunk 3 above)
Comment (ravil-mobile) [reply src/DynamicRupture/FrictionLaws/FrictionSolver.cpp:106]: auto-vectorization is done by a compiler

File: src/DynamicRupture/FrictionLaws/FrictionSolver.cpp (hunk 4 above)
Comment (ravil-mobile) [reply src/DynamicRupture/FrictionLaws/FrictionSolver.cpp:133]: resolved

**Review by sebwolf-de — APPROVED** [review]

**Pull request merged**